### ISSUE TITLE
Regarding concurrently persisting events

### ISSUE THREAD

**Issue by vach** [EventStore/EventStore#496]
Greg I was watching the 8hour video and got to the part of storing event in event store, where you talk about stored procedures and optimistic concurrency check... and again disruptor got into my mind... this is actually the exact problem it solves...

For example take a look at apache log4j2 they too had multiple concurrent producers (loggers) who push all logs into ringbuffer and that brought them 18x performance improvement over log4j 1.x... not even mentioning how latency improved, everything remains the same we dont change any piece of business logic that concurrently logs to log4j and suddenly we get 18x improvement because this data structure has no contention and no locks and batches things for us.

http://logging.apache.org/log4j/2.x/manual/async.html

 Imho it can be very valuable when using event store, not directly in eventstore but in any application that uses it, first of all it could use advantage of batching, when you get too many events to be persisted you could simply batch instert them without actually change the business logic that produced those events... and the second optimization would be that you could do the concurrency check before going to ES, so if you have a optimistic concurrency exception you can handle it before asking db to persist anything...

In short you never will have concurrent direct updates to event store from single application. 
What you think of it? is there something i'm missing?

**Comment by gregoryyoung** [comment]
This already happens. Requests are linearized both in an individual
connection and in event store itself

In terms of batching we do not batch at a client API level but do batch
internally in event store itself

In terms of batch inserting you would only be able to do this per stream so
it would be very limited in terms of its value. I would recommend doing
this yourself on top of the connection (and its quite easy to do).

Cheers,

Greg

On Mon, Jun 1, 2015 at 6:06 PM, Vachagan Balayan <notifications@github.com>
wrote:

> Greg I was watching the 8hour video and got to the part of storing event
> in event store, where you talk about stored procedures and optimistic
> concurrency check... and again disruptor got into my mind... this is
> actually the exact problem it solves...
>
> For example take a look at apache log4j2 they too had multiple concurrent
> producers (loggers) who push all logs into ringbuffer and that brought them
> 18x performance improvement over log4j 1.x... not even mentioning how
> latency improved, everything remains the same we dont change any piece of
> business logic that concurrently logs to log4j and suddenly we get 18x
> improvement because this data structure has no contention and no locks and
> batches things for us.
>
> http://logging.apache.org/log4j/2.x/manual/async.html
>
> Imho it can be very valuable when using event store, not directly in
> eventstore but in any application that uses it, first of all it could use
> advantage of batching, when you get too many events to be persisted you
> could simply batch instert them without actually change the business logic
> that produced those events... and the second optimization would be that you
> could do the concurrency check before going to ES, so if you have a
> optimistic concurrency exception you can handle it before asking db to
> persist anything...
>
> In short you never will have concurrent direct updates to event store from
> single application.
> What you think of it? is there something i'm missing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/EventStore/EventStore/issues/496>.
>



-- 
Studying for the Turing test

**Comment by vach** [comment]
I don't really understand how on eventstore side you batch? do you force it to wait say 10ms and watever is there batch insert? because if so it is like fixed latency cost...

**Comment by vach** [comment]
Regarding "batch inserting per stream" doesent all events go to main event stream? i imagine it is a single long stream which is then projected to other streams...

**Comment by gregoryyoung** [comment]
its dynamic and quite simple.

Basically there is a queue in front of the thread that is persisting to
disk. The rules are as follows.

if queue empty flush
If maxflushdelay has been hit flush

So if you do one write ... it writes, queue is empty and flushes
If you do 5. Write Write Write Write Write Flush providing max has not been
hit along the way.

Cheers,

Greg

On Mon, Jun 1, 2015 at 6:20 PM, Vachagan Balayan <notifications@github.com>
wrote:

> I don't really understand how on eventstore side you batch? do you force
> it to wait say 10ms and watever is there batch insert? because if so it is
> like fixed latency cost...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/EventStore/EventStore/issues/496#issuecomment-107576702>
> .
>



-- 
Studying for the Turing test

**Comment by gregoryyoung** [comment]
If you presume to only ever support a single replica set sure.

On Mon, Jun 1, 2015 at 6:22 PM, Vachagan Balayan <notifications@github.com>
wrote:

> Regarding "batch inserting per stream" doesent all events go to main event
> stream? i imagine it is a single long stream which is then projected to
> other streams...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/EventStore/EventStore/issues/496#issuecomment-107578328>
> .
>



-- 
Studying for the Turing test

**Comment by vach** [comment]
Thanks, learning a lot from your contributions.
